Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTree.java

```diff
@@ -0,0 +1,538 @@
+                return internalNodes.rightIndex[parent];
+            } else if (internalNodes.rightIndex[parent] == nodeOffset) {
+                return internalNodes.leftIndex[parent];
+            } else {
+                throw new IllegalArgumentException("node parent does not link back to node");
+            }
+        } else
+            return -1; // root is leaf
+    }
+
+    public boolean leftOf(float[] point, short nodeOffset) {
+        return leftOf(CommonUtils.toDoubleArray(point), nodeOffset);
+    }
+
+    public boolean leftOf(double[] point, short nodeOffset) {
+        checkArgument(!isLeaf(nodeOffset), " error ");
+        return (point[internalNodes.cutDimension[nodeOffset]] <= internalNodes.cutValue[nodeOffset]);
+    }
+
+    /**
+     * Delete the given point from this tree.
+     *
+     * @param sequential An offset of the point in the tree that we wish to delete.
+     */
+    /*
+     * public void deletePoint(WeightedPoint weightedPoint) { checkState(root !=
+     * null, "root must not be null"); deletePoint(root, weightedPoint.getPoint(),
+     * weightedPoint.getSequenceIndex()); }
+     */
+    @Override
+    public void deletePoint(Sequential<Integer> sequential) {
+        checkState(rootIndex != -1, "root must not be null");
+        deletePoint(rootIndex, pointStore.get(sequential.getValue()), 0);
+    }
+
+    /**
+     * This function deletes the point from the tree recursively. We traverse the
+     * tree based on the cut stored in each interior node until we reach a leaf
+     * node. We then delete the leaf node if the mass of the node is 1, otherwise we
+     * reduce the mass by 1.
+     *
+     * @param nodeOffset node that we are visiting in the tree.
+     * @param point      the point that is being deleted from the tree.
+     */
+    private int deletePoint(short nodeOffset, float[] point, int level) {
+
+        if (isLeaf(nodeOffset)) {
+            short leafOffset = (short) (nodeOffset - maxSize);
+            if (!pointStore.pointEquals(leafNodes.pointIndex[leafOffset], point)) {
+                throw new IllegalStateException(
+                        Arrays.toString(point) + " " + Arrays.toString(pointStore.get(leafNodes.pointIndex[leafOffset]))
+                                + " " + leafOffset + " node " + leafNodes.pointIndex[leafOffset] + " " + false
+                                + " Inconsistency in trees in delete step here.");
+            }
+
+            //
+            // the above assumes that sequence indexes are unique ... which is true for the
+            // specific sampler used
+            //
+
+            if (leafNodes.mass[leafOffset] > 1) {
+                --leafNodes.mass[leafOffset];
+                return leafNodes.pointIndex[leafOffset];
+            }
+
+            short parent = leafNodes.parentIndex[leafOffset];
+            int saved = leafNodes.pointIndex[leafOffset];
+            if (parent == -1) {
```

Comment:
minor. this check can also have a named method isEmpty/None for readability.

Reply:
I split the difference and defined a new constant
```
public final static short NULL = -1;
```
So now this looks like
```
if (parent == NULL)
```
I felt that was clearer than `isNone(parent)`, but let me know what you think.